### PR TITLE
Add `IfAny` and `IsAny` types

### PR DESCRIPTION
**Pull request by nullndr**
## Changelog

This PR adds two new types: `IfAny` and `IsAny`.

### IfAny

Given three types `Cond`, `Then` and `Else` assign `Then` if `Cond` is any, `Else` otherwise.

### IsAny

Returns `true` or `false` indicating if `Cond` is `any`.

## Links

For more info, see [this stack overflow answer](https://stackoverflow.com/questions/49927523/disallow-call-with-any/49928360#49928360).

## Notes

I'm drafting this PR because I don't know in what categories these types should be. 
I'm waiting for a response to this doubt.

### CODE REVIEW

**Comment by sindresorhus** [comment]
Seems like useful types.

**Comment by sindresorhus** [comment]
Please follow: https://github.com/sindresorhus/type-fest/blob/main/.github/contributing.md#submitting-a-new-type

**Comment by sindresorhus** [comment]
> I'm drafting this PR because I don't know in what categories these types should be.

Just put it in `Utilities` for now.

**Comment by sindresorhus** [comment]
You also need to add them to the readme.

**Comment by nullndr** [comment]
> You also need to add them to the readme.

Yes, I was waiting for the category, I'm gonna update this PR in 10 minutes

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.d.ts [hunk 1]
```diff
@@ -102,3 +102,6 @@ export type {LastArrayElement} from './source/last-array-element';
 // Miscellaneous
 export type {PackageJson} from './source/package-json';
 export type {TsConfigJson} from './source/tsconfig-json';
+
+export type {IfAny} from './source/if-any';
+export type {IsAny} from './source/is-any';
```
Comment (nullndr) [inline index.d.ts:107]: Oh sorry I'm missing these

**Comment by sindresorhus** [comment]
Try to look over your own diff. There a lots of minor things to improve. Basically, spend some more time perfecting it.

Also make sure you follow everything in https://github.com/sindresorhus/type-fest/blob/main/.github/contributing.md#submitting-a-new-type

**Comment by nullndr** [comment]
@sindresorhus sorry to bother you but could you tell me what's still missing in your opinion?

File: index.d.ts [hunk 2]
```diff
@@ -101,4 +103,4 @@ export type {LastArrayElement} from './source/last-array-element';
 
 // Miscellaneous
 export type {PackageJson} from './source/package-json';
-export type {TsConfigJson} from './source/tsconfig-json';
+export type {TsConfigJson} from './source/tsconfig-json';
```
Comment (sindresorhus) [inline index.d.ts:106]: Don't make unrelated changes.

File: test-d/if-any.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+import type {IfAny} from '../index';
+
+expectAssignable<IfAny<any, number, string>>(1);
+
+expectNotAssignable<IfAny<any, number, string>>('hello');
+
+expectAssignable<IfAny<any, string, number>>('world');
+
+expectNotAssignable<IfAny<any, string, number>>(2);
+
```
Comment (sindresorhus) [inline test-d/if-any.ts:11]:
```suggestion
```

File: source/is-any.d.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  if(should_print) {
+    console.log(data);
+  }
+}
+
+```
+
+@category Utilities
+ */
+export type IsAny<Cond> = 0 extends 1 & Cond ? true : false;
```
Comment (sindresorhus) [inline source/is-any.d.ts:20]: I don't think `Cond` is a correct name for this type parameter.

File: source/is-any.d.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+This type is useful when used inside other types.
+
+@example
+```ts
+import type {IsAny} from 'type-fest';
+
+function execute<T>(data: T, should_print: IsAny<T>) {
+  if(should_print) {
+    console.log(data);
+  }
```
Comment (sindresorhus) [inline source/is-any.d.ts:13]: Use tab-indentation

File: source/is-any.d.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+Returns a boolean indicating if the given type is any.
+
+This type is useful when used inside other types.
+
+@example
+```ts
+import type {IsAny} from 'type-fest';
+
+function execute<T>(data: T, should_print: IsAny<T>) {
+  if(should_print) {
```
Comment (sindresorhus) [inline source/is-any.d.ts:11]: Use proper formatting an camel-case.

File: source/if-any.d.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
+/**
+Used to create a ternary operator type above the `any` type.
```
Comment (sindresorhus) [inline source/if-any.d.ts:2]: This should be the same as the readme description.

File: source/if-any.d.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  c: string;
+  d: any;
+}
+
+type WithoutAnyValues = { [K in keyof WithAnyValues]: IfAny<WithAnyValues[K], number, WithAnyValues[K]>}
+
+let my_object: WithoutAnyValues = {
+  a: 1,
+  b: 2,
+  c: "hello",
```
Comment (sindresorhus) [inline source/if-any.d.ts:22]: single-quotes

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/type-fest/pull/505#discussion_r1004070861 is not resolved.

File: source/is-any.d.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+/**
+Returns a boolean indicating if the given type is any.
+
+This type is useful when used inside other types.
+
+@example
+```ts
+import type {IsAny} from 'type-fest';
+
+function execute<T>(data: T, should_print: IsAny<T>) {
```
Comment (sindresorhus) [inline source/is-any.d.ts:10]: It's not clear to the user why `IsAny` is used here.

File: source/if-any.d.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import type {IfAny} from 'type-fest';
+
+type WithAnyValues = {
+  a: any;
+  b: number;
+  c: string;
+  d: any;
+}
+
+type WithoutAnyValues = { [K in keyof WithAnyValues]: IfAny<WithAnyValues[K], number, WithAnyValues[K]>}
```
Comment (sindresorhus) [inline source/if-any.d.ts:17]: It would be better with a realistic real-world example.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/if-any.d.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
+/**
+Create a ternary operator type checking if the given type if `any`.
```
Comment (sindresorhus) [inline source/if-any.d.ts:2]: Typo: `if`

File: readme.md [hunk 12]
```diff
@@ -183,6 +183,8 @@ Click the type names for complete docs.
 - [`RequiredKeysOf`](source/required-keys-of.d.ts) - Extract all required keys from the given type.
 - [`HasRequiredKeys`](source/has-required-keys.d.ts) - Create a `true`/`false` type depending on whether the given type has any required fields.
 - [`Spread`](source/spread.d.ts) - Mimic the type inferred by TypeScript when merging two objects or two arrays/tuples using the spread syntax.
+- [`IfAny](source/if-any.d.ts) - Create a ternay operator type checking if the given type is `any`.
+- [`IsAny](source/is-any.d.ts) - Create a `true`/`false` type depending if the given type is `any`.
```
Comment (sindresorhus) [inline readme.md:187]:
```suggestion
- [`IsAny](source/is-any.d.ts) - Returns a `boolean` for whether the given type is `any`.
```

File: readme.md [hunk 13]
```diff
@@ -183,6 +183,8 @@ Click the type names for complete docs.
 - [`RequiredKeysOf`](source/required-keys-of.d.ts) - Extract all required keys from the given type.
 - [`HasRequiredKeys`](source/has-required-keys.d.ts) - Create a `true`/`false` type depending on whether the given type has any required fields.
 - [`Spread`](source/spread.d.ts) - Mimic the type inferred by TypeScript when merging two objects or two arrays/tuples using the spread syntax.
+- [`IfAny](source/if-any.d.ts) - Create a ternay operator type checking if the given type is `any`.
```
Comment (sindresorhus) [inline readme.md:186]: I think this needs a better description. It's not clear what "ternay operator type checking" means. Also typo.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
> Try to look over your own diff. There a lots of minor things to improve. Basically, spend some more time perfecting it.

Still applies.

**Comment by sindresorhus** [comment]
Bump :)

**Pull request closed without merging**